megatron/core/optimizer/__init__.py

```diff
@@ -48,100 +48,110 @@
-            satisfy scale_lr_cond.
-        lr (float): learning rate.
-        min_lr (float): minimum learning rate.
-        decoupled_lr (Optional[float]): optional decoupled learning rate.
-        decoupled_min_lr (Optional[float]): optional decoupled minimum learning rate.
-        default_skip_embedding_weight_decay (bool): whether to skip weight decay for embedding
-            parameters by default, if no_weight_decay_cond is not provided.
-
+        config (OptimizerConfig): optimizer configuration object.
+        config_overrides (Optional[Dict[LayerKey, OptimizerConfig]): optimizer overrides,
+            specified on a per-layer basis.
     Returns:
         List of parameter groups.
     """
 
-    use_decoupled_learning_rate = decoupled_lr is not None
-
-    # Map (wd_mult, lr_mult, is_expert_parallel, is_decoupled_lr) to params.
+    # Map (wd_mult, is_expert_parallel, param_group_hyperparameters_config) to params.
     params_map = {}
+    configs_map = {}
+
     for model_chunk in model_chunks:
         for name, param in model_chunk.named_parameters():
             if not param.requires_grad:
                 continue
 
+            # Get optimizer config for this parameter.
+            if config_overrides is None:
+                config_for_param = config
+            else:
+                config_for_param = None
+                for param_key in config_overrides:
+                    if _matches(param, name, param_key):
+                        config_for_param = config_overrides[param_key]
+                        break
+                # Fall back to default config.
+                if config_for_param is None:
+                    config_for_param = config
+
             is_expert_parallel = not getattr(param, 'allreduce', True)
 
-            if no_weight_decay_cond is not None:
-                no_wd: bool = no_weight_decay_cond(name, param)
-            else:
-                # Do not regularize biases and norm parameters.
-                #  optionally, also skip weight decay for embedding parameters if requested
-                #  (useful if you do not want embeddings to shrink to zero in training
-                #  https://arxiv.org/abs/2312.16903)
-                no_wd = (
-                    name.endswith(".bias")
-                    or len(param.shape) == 1
-                    or (default_skip_embedding_weight_decay and "embedding" in name)
-                )
+            # TODO: Figure out how to support backwards compatibility with param_groups
+            # with old keys (*, is_decoupled_lr).
 
-            if scale_lr_cond is not None:
-                scale_lr = scale_lr_cond(name, param)
+            # TODO: Make sure there is a way to support old no_weight_decay_func functionality
+            # and default_skip_embedding_weight_decay:
+            #     or (default_skip_embedding_weight_decay and "embedding" in name)
+            no_wd = name.endswith(".bias") or len(param.shape) == 1
```

Comment:
Should this be special functionality, or should we use a config override for this?